pkg/cephmgr/file/filesystem.go

```diff
@@ -0,0 +1,380 @@
+	key := path.Join(getMDSDesiredNodesKey(applied), fs.nodeID)
+	_, err := f.context.EtcdClient.Delete(ctx.Background(), key, &etcd.DeleteOptions{Dir: true, Recursive: true})
+	if err != nil {
+		return fmt.Errorf("failed to delete mds %s from desired state. %+v", fs.mdsID, err)
+	}
+
+	return nil
+}
+
+// Persist the file system config to the applied key
+func (f *FileSystem) markApplied() error {
+	return f.storeSettings(true)
+}
+
+// Remove the file system from the applied key
+func (f *FileSystem) markUnapplied() error {
+	_, err := f.context.EtcdClient.Delete(ctx.Background(), getFileKey(f.ID, true), &etcd.DeleteOptions{Dir: true, Recursive: true})
+	return err
+}
+
+func getFileRootKey(applied bool) string {
+	return path.Join(mon.CephKey, "file", getAppliedStr(applied))
+}
+
+func getFileKey(name string, applied bool) string {
+	return path.Join(mon.CephKey, "file", getAppliedStr(applied), name)
+}
+
+func getAppliedStr(applied bool) string {
+	if applied {
+		return clusterd.AppliedKey
+	}
+	return clusterd.DesiredKey
+}
+
+func (f *FileSystem) storeSettings(applied bool) error {
+
+	key := path.Join(getFileKey(f.ID, applied), poolKeyName)
+	_, err := f.context.EtcdClient.Set(ctx.Background(), key, f.Pool, nil)
+	return err
+}
+
+func (f *FileSystem) startMDSUnitInstances(count int) error {
+	desiredMDS, err := f.getDesiredMDSIDs(count)
+	if err != nil {
+		return fmt.Errorf("failed to get desired mds instances. %+v", err)
+	}
+
+	var nodeIDs []string
+	for _, mds := range desiredMDS {
+		nodeIDs = append(nodeIDs, mds.nodeID)
+	}
+
+	// trigger the monitors to start on each node
+	logger.Infof("Triggering mds on nodes: %+v", nodeIDs)
+	err = clusterd.TriggerAgentsAndWaitForCompletion(f.context.EtcdClient, nodeIDs, mdsAgentName, len(desiredMDS))
+	if err != nil {
+		return fmt.Errorf("failed to deploy mds agents. %+v", err)
+	}
+
+	// set the mds as applied
+	for _, mds := range desiredMDS {
+		if err := f.storeMDSState(mds, true); err != nil {
+			return fmt.Errorf("failed to set mds agent as applied. %+v", err)
+		}
+	}
+
+	return nil
+}
+
+func (f *FileSystem) stopMDSUnitInstances() error {
+	appliedMDS, _, _, err := f.loadMDS(f.ID, true)
+	if err != nil {
+		return fmt.Errorf("failed to get desired mds instances. %+v", err)
+	}
+
+	var nodeIDs []string
+	for _, mds := range appliedMDS {
+		nodeIDs = append(nodeIDs, mds.nodeID)
+	}
+
+	// trigger the monitors to start on each node
+	logger.Infof("Triggering removal of mds on nodes: %+v", nodeIDs)
+	err = clusterd.TriggerAgentsAndWaitForCompletion(f.context.EtcdClient, nodeIDs, mdsAgentName, len(nodeIDs))
+	if err != nil {
+		return fmt.Errorf("failed to deploy mds agents. %+v", err)
+	}
+
+	// remove the mds from applied
+	for _, mds := range appliedMDS {
+		if err := f.removeMDSState(mds, true); err != nil {
+			return fmt.Errorf("failed to set mds agent as applied. %+v", err)
+		}
+	}
+
+	return nil
+}
+
+func (f *FileSystem) getDesiredMDSIDs(count int) (map[string]*mdsInfo, error) {
```

Comment:
Acronyms are actually not recommended for camel case. I agree it's odd when you have two acronyms next to each other though. https://github.com/golang/go/wiki/CodeReviewComments#initialisms